plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java

```diff
@@ -175,50 +197,81 @@ private String determinePageLocation(IServer server, ILaunchConfiguration config
     return "http://" + DEBUGGER_HOST + ":8080";
   }
 
+
   /**
-   * Open a browser on the provided page on server start.
+   * Monitors the server and launch state. Ensures listeners are properly removed on server stop and
+   * launch termination. It is necessary in part as there may be several launches per
+   * LaunchConfigurationDelegate.
+   * <ul>
+   * <li>On server start, open a browser page</li>
+   * <li>On launch-termination, stop the server.</li>
+   * <li>On server stop, terminate the launch</li>
+   * </ul>
    */
-  private class OpenBrowserListener implements IServerListener {
-    private String pageLocation;
+  private class ServerLaunchMonitor implements ILaunchesListener2, IServerListener {
+    private ILaunchConfiguration configuration;
+    private ILaunch launch;
+    private IServer server;
+
+    /**
+     * Setup the monitor.
+     */
+    ServerLaunchMonitor(ILaunchConfiguration configuration, ILaunch launch, IServer server) {
+      this.configuration = configuration;
+      this.launch = launch;
+      this.server = server;
+    }
 
-    public OpenBrowserListener(String pageLocation) {
-      this.pageLocation = pageLocation;
+    private void engage() {
+      getLaunchManager().addLaunchListener(this);
+      server.addServerListener(this);
+    }
+
+    private void disengage() {
+      getLaunchManager().removeLaunchListener(this);
+      server.removeServerListener(this);
     }
 
     @Override
     public void serverChanged(ServerEvent event) {
-      if (event.getState() != IServer.STATE_STARTED) {
-        return;
-      }
-      final IServer server = event.getServer();
-      final IWorkbench workbench = PlatformUI.getWorkbench();
+      assert server == event.getServer();
```

Comment:
Ok.  Is there a rationale?  I suppose it's because nobody runs with `-enableassertions` and so they're never caught…